tools/kata-agent-ctl/src/client.rs

```diff
@@ -0,0 +1,1193 @@
+        fp: builtin_cmd_list,
+    },
+    BuiltinCmd {
+        name: "list",
+        descr: "List all available commands",
+        fp: builtin_cmd_list,
+    },
+    BuiltinCmd {
+        name: "repeat",
+        descr: "Repeat the next command 'n' times [-1 for forever]",
+        fp: builtin_cmd_repeat,
+    },
+    BuiltinCmd {
+        name: "sleep",
+        descr:
+            "Pause for specified period number of nanoseconds (supports human-readable suffixes [no floating points numbers])",
+        fp: builtin_cmd_sleep,
+    },
+    BuiltinCmd {
+        name: CMD_QUIT,
+        descr: "Exit this program",
+        fp: builtin_cmd_quit,
+    },
+];
+
+fn get_agent_cmd_names() -> Vec<String> {
+    let mut names = Vec::new();
+
+    for cmd in AGENT_CMDS {
+        names.push(cmd.name.to_string());
+    }
+
+    names
+}
+
+fn get_agent_cmd_details() -> Vec<String> {
+    let mut cmds = Vec::new();
+
+    for cmd in AGENT_CMDS {
+        let service = match cmd.st {
+            ServiceType::Agent => "agent",
+            ServiceType::Health => "health",
+        };
+
+        cmds.push(format!("{} ({} service)", cmd.name, service));
+    }
+
+    cmds
+}
+
+fn get_agent_cmd_func(name: &str) -> Result<AgentCmdFp> {
+    for cmd in AGENT_CMDS {
+        if cmd.name == name {
+            return Ok(cmd.fp);
+        }
+    }
+
+    Err(format!("Invalid command: {:?}", name))
+}
+
+fn get_builtin_cmd_details() -> Vec<String> {
+    let mut cmds = Vec::new();
+
+    for cmd in BUILTIN_CMDS {
+        cmds.push(format!("{} ({})", cmd.name, cmd.descr));
+    }
+
+    cmds
+}
+
+fn get_all_cmd_details() -> Vec<String> {
+    let mut cmds = get_builtin_cmd_details();
+
+    cmds.append(&mut get_agent_cmd_names());
+
+    cmds
+}
+
+fn get_builtin_cmd_func(name: &str) -> Result<BuiltinCmdFp> {
+    for cmd in BUILTIN_CMDS {
+        if cmd.name == name {
+            return Ok(cmd.fp);
+        }
+    }
+
+    Err(format!("Invalid command: {:?}", name))
+}
+
+// The vsock crate doesn't work ;((
+fn client_create_vsock_fd_with_nix_crate(cid: libc::c_uint, port: u32) -> Result<RawFd> {
+    info!(sl!(), "XXX: using NIX crate for client VSOCK comms");
+
+    let fd = match socket(
+        AddressFamily::Vsock,
+        SockType::Stream,
+        SockFlag::SOCK_CLOEXEC,
+        None,
+    ) {
+        Ok(fd) => fd,
+        Err(e) => return Err(e.to_string()),
+    };
+
+    let sock_addr = SockAddr::new_vsock(cid, port);
+
+    match connect(fd, &sock_addr) {
+        Ok(_) => (),
+        Err(e) => return Err(e.to_string()),
+    };
+
+    Ok(fd)
+}
+
+fn create_ttrpc_client(cid: libc::c_uint, port: u32) -> Result<ttrpc::Client> {
+    let fd = match client_create_vsock_fd_with_nix_crate(cid, port) {
+        Ok(fd) => fd,
+        Err(e) => {
+            return Err(format!(
+                "failed to create VSOCK connection (check agent is running): {:?}",
+                e
+            ))
+        }
+    };
```

Comment:
You could also use the crate `anyhow` so that each `Error` can be generated from a string. This would avoid the need for `pub type Result<T> = std::result::Result<T, String>;`, which is a special Result returning a string as an error.

Reply:
Done.